File: sdk/dart/lib/src/collection.dart
import 'dart:convert';

import 'http_client.dart';

class Collection {
  final BakendHttpClient http;
  final String name;

  Collection({required this.http, required this.name});

  Future<List<Map<String, dynamic>>> list() async {
    final body = await http.request<Map<String, dynamic>>('/api/$name');
    final items = body['items'] as List<dynamic>;
    return items.map((item) => item as Map<String, dynamic>).toList();
  }

  Future<Map<String, dynamic>> get(String id) async {
    return http.request<Map<String, dynamic>>('/api/$name/$id');
  }

  Future<Map<String, dynamic>> create(Map<String, dynamic> data) async {
    return http.request<Map<String, dynamic>>(
      '/api/$name',
      method: 'POST',
      body: jsonEncode(data),
    );
  }

  Future<Map<String, dynamic>> update(String id, Map<String, dynamic> data) async {
    return http.request<Map<String, dynamic>>(
      '/api/$name/$id',
      method: 'PUT',
      body: jsonEncode(data),
    );
  }

  Future<void> delete(String id) async {
    await http.request<void>(
      '/api/$name/$id',
      method: 'DELETE',
    );
  }
}

File: sdk/javascript/src/collection.ts
import type { HttpClient, ListResult, RecordData } from "./types.ts";

export class Collection<T extends RecordData = RecordData> {
  private readonly http: HttpClient;
  private readonly name: string;

  constructor(http: HttpClient, name: string) {
    this.http = http;
    this.name = name;
  }

  async list(): Promise<T[]> {
    const body = await this.http.request<ListResult<T>>(`/api/${this.name}`);
    return body.items;
  }

  async get(id: string): Promise<T> {
    return this.http.request<T>(`/api/${this.name}/${id}`);
  }

  async create(data: Partial<T>): Promise<T> {
    return this.http.request<T>(`/api/${this.name}`, {
      method: "POST",
      body: JSON.stringify(data),
    });
  }

  async update(id: string, data: Partial<T>): Promise<T> {
    return this.http.request<T>(`/api/${this.name}/${id}`, {
      method: "PUT",
      body: JSON.stringify(data),
    });
  }

  async delete(id: string): Promise<void> {
    await this.http.request<void>(`/api/${this.name}/${id}`, {
      method: "DELETE",
    });
  }
}
